aesd-char-driver/main.c

```diff
@@ -0,0 +1,259 @@
+        {
+            // Copy data to user space
+            bytes_to_copy = ( available_bytes > count ) ? count : available_bytes;
+
+            if ( !copy_to_user( buf, entry->buffptr + entry_offset_byte, bytes_to_copy ))
+            {
+                // Update file position
+                *f_pos += bytes_to_copy;
+                retval = bytes_to_copy;
+            }
+            else
+            {
+                // Error copying data to user space
+                retval = -EFAULT;
+                goto errorHandler;
+            }
+        }
+        else
+        {
+            // No more data available
+            retval = 0;
+            goto errorHandler;
+        }
+    }
+    else
+    {
+        // No data available here
+        retval = 0;
+        goto errorHandler;
+    }
+
+errorHandler:
+    // Unlock the device
+    mutex_unlock( &aesdDev->mutexLock );
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = -ENOMEM;
+    struct aesd_dev *aesdDev = NULL;
+    const char* entryPtr = NULL;
+    // PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    PDEBUG("write %zu bytes with offset %lld: %.*s", count, *f_pos, (int)count, buf);
+
+    aesdDev = filp->private_data;
+
+    // Allocate memory for the write command
+    aesdDev->buffEntry.buffptr = aesdDev->buffEntry.size ? ( char * )krealloc( aesdDev->buffEntry.buffptr, count + aesdDev->buffEntry.size, GFP_KERNEL ) : ( char * )kmalloc( count, GFP_KERNEL );
+
+    if ( aesdDev->buffEntry.buffptr == NULL )
+    {
+        // Error handling
+        return -ENOMEM;
+    }
+
+    if ( mutex_lock_interruptible( &aesd_device.mutexLock ) )
+    {
+        // Unable to lock mutex
+        retval = -ERESTARTSYS;
+    }
+
+    // Copy write command from user space
+    if ( copy_from_user( (void *)( aesdDev->buffEntry.buffptr + aesdDev->buffEntry.size ), buf, count ) == 0 )
+    {
+        aesdDev->buffEntry.size += count;
+        if ( memchr( aesdDev->buffEntry.buffptr, '\n', aesdDev->buffEntry.size ) != NULL )
+        {
+            // Add the entry to the circular buffer
+            entryPtr = aesd_circular_buffer_add_entry( &aesdDev->buff, &aesdDev->buffEntry );
+
+            if ( entryPtr != NULL )
+            {
+                kfree( entryPtr );
+                entryPtr = NULL;
+            }
+
+            aesdDev->buffEntry.buffptr = NULL;
+            aesdDev->buffEntry.size = 0;
+        }
+    }
+    else
+    {
+        kfree( aesdDev->buffEntry.buffptr );
+        // Error copying from user space
+        retval = -EFAULT;
+        goto errorHandler;
+    }
+
+    // Unlock the device
+    mutex_unlock( &aesd_device.mutexLock );
+
+    retval = count; // Return the number of bytes written
+
+    return retval;
+
+errorHandler:
+    // Free memory allocated for buffer entry
+    kfree( aesdDev->buffEntry.buffptr );
```

Comment:
@Kanin-McGuire will have a double free here I think as currently implemented.  I think you could just remove this goto label completely and fall through in the else case above.